### ISSUE TITLE
APIServer panic because of concurrent map iteration and map write

### ISSUE THREAD

**Issue by NickrenREN** [kubernetes/kubernetes#80449]
**What happened**:

apiserver panic because of concurrent map iteration and map write
but it happens occasionally， can not reproduce it easily.

**What you expected to happen**:

will not panic

**How to reproduce it (as minimally and precisely as possible)**:

It happens occasionally， can not reproduce it easily.

**Anything else we need to know?**:

Kubernetes version: 1.10.11
The kubernetes version is quite old, i am not sure if it is fixed in the newer versions

Part of logs are as follows
```
Jul 19 10:50:37 n225-068-092 hyperkube[629498]: [restful] 2019/07/19 10:50:37 log.go:33: [restful/swagger] listing is available at https://***:6443/swaggerapi
Jul 19 10:50:37 n225-068-092 hyperkube[629498]: [restful] 2019/07/19 10:50:37 log.go:33: [restful/swagger] https://***:6443/swaggerui/ is mapped to folder /swagger-ui/
Jul 19 10:50:39 n225-068-092 hyperkube[629498]: [restful] 2019/07/19 10:50:39 log.go:33: [restful/swagger] listing is available at https://***:6443/swaggerapi
Jul 19 10:50:39 n225-068-092 hyperkube[629498]: [restful] 2019/07/19 10:50:39 log.go:33: [restful/swagger] https://***:6443/swaggerui/ is mapped to folder /swagger-ui/
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: fatal error: concurrent map iteration and map write
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: goroutine 23218 [running]:
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: runtime.throw(0x4abe3b5, 0x26)
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: /usr/local/go/src/runtime/panic.go:608 +0x72 fp=0xc0ca9ced88 sp=0xc0ca9ced58 pc=0x42f632
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: runtime.mapiternext(0xc0ca9cef90)
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: /usr/local/go/src/runtime/map.go:790 +0x525 fp=0xc0ca9cee10 sp=0xc0ca9ced88 pc=0x411e15
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1.(*ObjectMeta).Size(0xc0ca9cf100, 0x2)
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: /opt/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/generated.pb.go:2053 +0x5a7 fp=0xc0ca9cf060 sp=0xc0ca9cee10 pc=0x88a4e7
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: k8s.io/kubernetes/vendor/k8s.io/api/apps/v1.(*Deployment).Size(0xc0ca9cf0e0, 0x2)
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: /opt/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/api/apps/v1/generated.pb.go:1594 +0x35 fp=0xc0ca9cf0a8 sp=0xc0ca9cf060 pc=0xc0b1f5
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: k8s.io/kubernetes/vendor/k8s.io/api/apps/v1.(*DeploymentList).Size(0xc0cc075b20, 0x7b54420)
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: /opt/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/api/apps/v1/generated.pb.go:1628 +0x118 fp=0xc0ca9cf8d0 sp=0xc0ca9cf0a8 pc=0xc0b558
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/protobuf.(*Serializer).Encode(0xc000154550, 0x7b54420, 0xc0cc075b20, 0x7f10d5fd1dc0, 0xc097964b00, 0x7b54400, 0xc0cc075b20)
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: /opt/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf.go:200 +0x2bf fp=0xc0ca9cfab0 sp=0xc0ca9cf8d0 pc=0xa3d8ff
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/versioning.(*codec).Encode(0xc0cc9ae080, 0x7b4d320, 0xc0e27f6540, 0x7f10d5fd1dc0, 0xc097964b00, 0xc0d1f05240, 0x7f118fd35ae0)
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: /opt/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go:208 +0x562 fp=0xc0ca9cfc30 sp=0xc0ca9cfab0 pc=0xa3b3d2
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.SerializeObject(0x4aac30f, 0x23, 0x7f118fd35ae0, 0xc0cc9ae080, 0x7b79620, 0xc097964b00, 0xc0e2832500, 0xc8, 0x7b4d320, 0xc0e27f6540)
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: /opt/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:97 +0xe9 fp=0xc0ca9cfc80 sp=0xc0ca9cfc30 pc=0x11a74d9
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.WriteObjectNegotiated(0x7f118fdbd898, 0xc0e35ea300, 0x7b7d760, 0xc0030332c0, 0x4a1fae8, 0x4, 0x4a1e8c1, 0x2, 0x7b79620, 0xc097964b00, ...)
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: /opt/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:123 +0x361 fp=0xc0ca9cfdb0 sp=0xc0ca9cfc80 pc=0x11a78a1
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.WriteObject(0x7f118fdbd898, 0xc0e35ea300, 0xc8, 0x4a1fae8, 0x4, 0x4a1e8c1, 0x2, 0x7b7d760, 0xc0030332c0, 0x7b4d320, ...)
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: /opt/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:52 +0x29a fp=0xc0ca9cfeb8 sp=0xc0ca9cfdb0 pc=0x11a6c4a
Jul 19 10:50:44 n225-068-092 hyperkube[629498]: k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.transformResponseObject(0x7f118fdbd898, 0xc0e35ea300, 0x7b9e900, 0xc009a4
... 
```

**Comment by NickrenREN** [comment]
@kubernetes/sig-api-machinery-bugs

**Comment by NickrenREN** [comment]
It seems we run into the panic when we try to iterate the deployment's annotation

**Comment by NickrenREN** [comment]
It seems that we run into the panic when we iterate the deployment's annotation map

**Comment by yue9944882** [comment]
> Kubernetes version: 1.10.11

are you using the exact upstream release or a customized fork? fwiw releases before 1.12 are no longer supported.

**Comment by NickrenREN** [comment]
@yue9944882  Based on upstream release 1.10.11, and we customize a little, we maintain it ourselves.

**Comment by liggitt** [comment]
I thought this was a duplicate of https://github.com/kubernetes/kubernetes/pull/65092, fixed in 1.11, but this appears to be involving deployments and proto Size() serialization, which is not the same

**Comment by liggitt** [comment]
> @yue9944882 Based on upstream release 1.10.11, and we customize a little, we maintain it ourselves.

can you provide the diff of the customizations against the official `v1.10.11`? do you pull in a different version of the proto libraries, or make any API changes?

**Comment by liggitt** [comment]
since this hasn't been reported against official versions, marking as support for now

**Comment by NickrenREN** [comment]
> can you provide the diff of the customizations against the official v1.10.11? do you pull in a different version of the proto libraries, or make any API changes?

@liggitt  We customize the code mostly on scheduler and controller parts, and do not make much changes on api and the proto libraries.

**Comment by liggitt** [comment]
> @liggitt We customize the code mostly on scheduler and controller parts, and do not make much changes on api and the proto libraries.

The line in the panic (`k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/generated.pb.go:2053`) does not match a map access line on v1.10.11:

https://github.com/kubernetes/kubernetes/blob/v1.10.11/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/generated.pb.go#L2053

Do you have the diff for that file in particular between your build and the v1.10.11 tag?

**Comment by ricky1993** [comment]
> > @liggitt We customize the code mostly on scheduler and controller parts, and do not make much changes on api and the proto libraries.
> 
> The line in the panic (`k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/generated.pb.go:2053`) does not match a map access line on v1.10.11:
> 
> https://github.com/kubernetes/kubernetes/blob/v1.10.11/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/generated.pb.go#L2053
> 
> Do you have the diff for that file in particular between your build and the v1.10.11 tag?

```
--- a/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/generated.pb.go
+++ b/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/generated.pb.go
@@ -1,5 +1,5 @@
 /*
-Copyright 2018 The Kubernetes Authors.
+Copyright 2019 The Kubernetes Authors.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
@@ -1111,6 +1111,14 @@ func (m *ListOptions) MarshalTo(dAtA []byte) (int, error) {
        i++
        i = encodeVarintGenerated(dAtA, i, uint64(len(m.Continue)))
        i += copy(dAtA[i:], m.Continue)
+       dAtA[i] = 0x48
+       i++
+       if m.AllowWatchBookmarks {
+               dAtA[i] = 1
+       } else {
+               dAtA[i] = 0
+       }
+       i++
        return i, nil
 }

@@ -2004,6 +2012,7 @@ func (m *ListOptions) Size() (n int) {
        n += 1 + sovGenerated(uint64(m.Limit))
        l = len(m.Continue)
        n += 1 + l + sovGenerated(uint64(l))
+       n += 2
        return n
 }

@@ -2430,6 +2439,7 @@ func (this *ListOptions) String() string {
                `IncludeUninitialized:` + fmt.Sprintf("%v", this.IncludeUninitialized) + `,`,
                `Limit:` + fmt.Sprintf("%v", this.Limit) + `,`,
                `Continue:` + fmt.Sprintf("%v", this.Continue) + `,`,
+               `AllowWatchBookmarks:` + fmt.Sprintf("%v", this.AllowWatchBookmarks) + `,`,
                `}`,
        }, "")
        return s
```
It seems hit the line here.
https://github.com/kubernetes/kubernetes/blob/637c7e288581ee40ab4ca210618a89a555b6e7e9/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/generated.pb.go#L2044

**Comment by liggitt** [comment]
Ah, it is the modification of the Annotations map in conversion. That was fixed in 1.11+ in https://github.com/kubernetes/kubernetes/pull/64267

See where it touched the deployments object (which is in this issue's panic stack) at https://github.com/kubernetes/kubernetes/pull/64267/files#diff-2ee4fc301db0484dfdcdc19887921873

/close
